### PR TITLE
ESM support

### PR DESCRIPTION
**Pull request by bkaney**
First pass to get both require and include (node and ESM) working.

Examples use node/require. Tests use ESM, but needed to follow a bit of a filename hack as described here: https://alxgbsn.co.uk/2019/02/22/testing-native-es-modules-mocha-esm/

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -8,7 +8,7 @@ install:
   - yarn install
 
 script:
-  - yarn test
+  - yarn test --require esm
```
Comment (foxtree) [inline .travis.yml]: this could be reverted, now that ` --require esm` is in pkg.scripts.test ?

**Comment by bkaney** [comment]
@foxtree - we’ll need to figure out how to get code coverage to work here.

**Comment by vstrateman** [comment]
> @foxtree - we’ll need to figure out how to get code coverage to work here.

Could it be related to this?
https://github.com/istanbuljs/nyc/issues/1131

**Comment by awatson1978** [comment]
Hmmm...  getting the following:

```
Unable to resolve some modules:

  "internal/bootstrap/loaders" in             
/Volumes/StorageCrystal-Quartz/Code/ClinicalFramework/material-fhir-demo/node_modules/fhir-kit-client/node_modules/esm/esm.js
(web.browser)

If you notice problems related to these missing modules, consider running:
                                              
  meteor npm install --save internal   
```

I've tried installing the `internal` package, but with no success.

**Comment by bkaney** [comment]
Yea, we are seeing a similar problem with React Native (we got it to work without problem in a plain React app using import). We will continue to investigate this. Thanks for your feedback!

**Comment by bkaney** [comment]
@awatson1978 -- can you try now (I moved esm to the dependencies, it was in devDeps)? And/or do you have a sample project that you could push somewhere so I might be able to iterate on it?

**Comment by bkaney** [comment]
The problem with react-native is support for .mjs files (https://github.com/facebook/metro/pull/60) -- maybe it's time to see if this could get advanced?

**Comment by bkaney** [comment]
See also: https://github.com/facebook/metro/pull/307

**Comment by awatson1978** [comment]
Hey!  Happy to report that I got it to work.  :)

I needed to change line [Line 15](https://github.com/Vermonster/fhir-kit-client/blob/esm-support/package.json#L15) in `package.json` to the following, with a normal .js extension:
```
  "main": "lib/client.js"
```

**Comment by awatson1978** [comment]
Code sample:

```js
import Client from 'fhir-kit-client';
const fhirClient = new Client({
  baseUrl: 'https://epicproxy.rush.edu/fhir-prd/api/FHIR/DSTU2/'
});

onButtonClick = function(){
  console.log('onButtonClick().fhirClient.smartAuthMetadata()', fhirClient.smartAuthMetadata())
  fhirClient.smartAuthMetadata().then((response) => {
    console.log(response);
  });
}
const App = () => (
  <div>
    <RaisedButton label="Fhir!" onClick={onButtonClick.bind(this)} /><br />
  </div>
);
```

**Comment by bkaney** [comment]
> I needed to change line [Line 15](https://github.com/Vermonster/fhir-kit-client/blob/esm-support/package.json#L15) in `package.json` to the following, with a normal .js extension:

Hmm.. that will break node/CommonJS users. Since you are using ESM, I was hoping your build system would use the `module` declaration in `package.json`. See https://rollupjs.org/guide/en#compatibility -- "If your `package.json` file also has a `module` field, ESM-aware tools like Rollup and webpack 2+ will import the ES module version directly."

**Comment by bkaney** [comment]
@awatson1978 are you using Metro to deal with JS bundling in your project? (https://github.com/facebook/metro)

**Comment by awatson1978** [comment]
Hi, sorry about the delay.  Had Q2 publication to work on and then 4th of July.  Yeah, the Meteor build pipeline doesn't seem to look at the 'module' field in the packages.  It should be configurable with a .bablerc file to override the defaults, but I've not found the correct syntax yet.  

Thank you for the `meteor-client-bundler` lead.  Will see if I can get anything to work using that.

**Comment by bkaney** [comment]
> Thank you for the `meteor-client-bundler` lead. Will see if I can get anything to work using that.

Hi @awatson1978 -- I am thinking we will follow this pattern: https://github.com/nodejs/modules/issues/273

```json
  "type": "module",
  "main": "index.cjs",
  "module": "index.js"
```

Then write in the REAMDE something like was commented in that thread (e.g. what you would do):

"all you have to write in your README is import stuff from 'index.js' in case bundlers are not used or are incapable of solving". So if your bundler doesn't know about `module`, you can do something like this:

```js
import Client from 'fhir-kit-client/index.js';
```

**Pull request closed without merging**